### ISSUE TITLE
Efficient Estimation of Word Representations in Vector Space

### ISSUE THREAD

**Issue by chullhwan-song** [chullhwan-song/Reading-Paper#36]
https://arxiv.org/abs/1301.3781

**Comment by chullhwan-song** [comment]
### Abstract
* 연속적인  단어에 대한 재표현할수 있는벡터을 계산할수 있는 2가지 새로운 방법을 제시
* 이들 재표현된 벡터의 퀄리티의 측정방법은 word 유사성 task에서
* 이 방법으로 이전다른 방법과 비교하여 best performance 
* 학습속도도 빠르게
* 자기네들이 최신 데이터셋도 제공
### Introduction
* 현재 많은 NLP 시스템 및 기술들은
    * word를 하나의 작은 단위로 취급
    * voc의 index로 치환
    * word사이의 유사성관련 notion은 들어가지 않음.
    * 위의 선택하는 이유는 simplicity, robustness, observation
        * simple models trained on huge amounts of data outperform complex systems trained on less data. 
    * 이러한 대표적 방법은 N-gram이다. (통계학적인 언어 모델링을 위해 많이 사용)  
* 위의 설명과 같이 대용량의 데이터로 가지고 단순한 모델로 학습한 모델이 위의 제시된 모델보다 훨씬 좋음.     
* 이 논문에서,     
    * 수십억의 단어와 어휘집에 존재하는 수백만의 단어들로 이루어진 매우 많은 데이터로 부터 높은 수준의 단어 벡터 학습을 하는 기술 소개
    * 어떤 이전의 논문들도 단어 벡터의 차원을 보통의 50- 100 차원으로하며, 수백만의 단어보다 많은 데이터를 이용해 성공적으로 훈련하지 못했음.
    * 가까운 (비슷한) 단어들의 근접성 & 단어가 다수의 유사도(multiple degrees of similarity)를 갖도록함.
       * 굴절어(라틴어 등등)의 문맥에서 먼저 관측
       * 예를 들어, 유사한 단어를 찾을때, 명사는 다양한 어미를 가지는데, 비슷한 어미를 갖는 단어들이 나올수 있다. 이게 유사한 단어인가?
       * 제시한 방법은 이러한 문제점을 뛰어넘는 '유사도를 발견할수 있음.
       * 단어 간격 기술을 이용해 단어 벡터의 대수적 연산(algebraic operations are performed on the word vectors)을 실행 >  ‘King’ - ‘Man’ + ‘Woman’ = ‘Queen’ 
### Model Architectures
* "continuous representations of word"에 대한 이전 잘알려진 방법중에는 Latent Semantic Analysis (LSA) ,Latent Dirichlet Allocation (LDA) 있음. 
    * 이런 모델은 topic model이라고 해서, 이전 딥러닝 이전에 핫하게 연구되고 있었음.  doc 이나 image안에 특정 latent를 예측하는 모델, 즉 topic 이라는 개념을 도입
    * 하지만, 이런 모델은 very expensive
* 따라서 우리는 "Strategies for Training Large Scale Neural Network Language Models" 연구와 비슷하게,  모든 학습된 모델에 대해 접근가능토록 "the number of parameters"를 계산하도록 정의한다.
   * computational complexity가 최소화하도록 시도하고, accuracy를 max가 되도록 한다.
   * 그래서,  the training complexity를 다음과 같이 정의
![image](https://user-images.githubusercontent.com/40360823/43691881-5f9b82ec-995d-11e8-9c1a-e2a0a8e0814f.png)
       * E는 epochs, T는 training set의 word 수, Q는 모델 구조
       *  epochs = 3`50,  T는 1 billon
       * 모든 모델은 stochastic gradient descent & backpropagation 이용하여 학습
### Model Architectures - Feedforward Neural Net Language Model(NNLM)
* feedforward neural network는 "A neural probabilistic language model. Journal of Machine Learning Research"에서 제안
![image](https://user-images.githubusercontent.com/40360823/43691875-4b32051a-995d-11e8-8daa-bd9c5226949a.png)
![image](https://user-images.githubusercontent.com/40360823/43691887-6b890c46-995d-11e8-8d1d-1b5b8d22affd.png)
* input = N , projection = P, hidden = H, output 구조
   * input layer는, 
      * N previous words = 1-of-V coding를 이용하여 encoding, V는 voc size > 1-of-V coding ? 
         * = 단어 이전의 단어들 N개를 one-hot encoding 으로 벡터화
  * N은 공유된 projection matrix를 이용하여 NxD 차원을 가진  P layer에 project된다. > D ??
      * P layer의 구성은 cheap operation
      * P layer는 그냥 mapping table matrix 아닌가??
  *  N =10개로 선택하면, H layer는 500-1000개
      * **the hidden layer is used to compute probability distribution over all the words in the
vocabulary**
      * 그래서 이때의 complexity는
![image](https://media.oss.navercorp.com/user/1196/files/4c508ed6-dea4-11e7-8631-c49b72d9dc15)
          * 가장 큰부분은 HxV
             * 이를 회피하기 위해, hierarchical softmax 적용하면, log2(V)로 작아짐.
          * 그래서, 가장 큰 복잡도는 NxDxH가 된다.
  * 어휘가 Huffman 이진 트리로 표현되고 계층적 softmax를 사용 - 속도 빠름.
### Model Architectures - Recurrent Neural Net Language Model(RNNLM)
* RNN 적용
* 위의에 언급한 NNLM의 한계를 극복하기 위해 제안.
* Projection Layer 없이 Input, Hidden, Output Layer로만 구성

**Comment by chullhwan-song** [comment]
* Noise-Contrastive 트레이닝를 이용한 Scaling Up
  * 이전의 히스토리 h (“history”)가 주어졌을때 다음에 이어지는 목표 단어 w_t (“target”)의 확률을 최대화하기 위해서 소프트맥스 함수를 사용한다.
![image](https://user-images.githubusercontent.com/40360823/43691895-840926f2-995d-11e8-9b21-3fbdc7462d15.png)
![image](https://user-images.githubusercontent.com/40360823/43691901-9c550a28-995d-11e8-9d6b-d9be4f6a6637.png)현재 콘텍스트 h에서 단어 w_t의 적합성을 나타낸다. (보통 내적을 이용한다.) 우리는 학습셋에서 log-likelihood를 최대화하는 방향으로 학습을 진행, 즉, 아래의 식을 최대화
   ![image](https://user-images.githubusercontent.com/40360823/43691906-a2ee223e-995d-11e8-99a4-4387ae6fdc3a.png)

**Comment by chullhwan-song** [comment]
### dict 크기 즉, 단어가 아주 많을때는 어떻게 처리하는 방법이 따로 있는가? 

 *  CBOW와 Skip-gram 모델을 그대로 구현해서 돌릴 경우 생각보다 학습이 오래 걸리는 것을 확인할 수 있을 것
  * 즉, 사전의 크기가 엄청나게 크면 문제가 생김.
  * 디테일하게 가면, Output Layer에서 Softmax 계산를 위해 전체 단어를 위해 normalized 계산을 하게되어 계산량이 엄청나게 많아짐.
   * 이런한 방법을 해결하기 위한 방법이 바로 Hierarchical Softmax vs Negative Sampling

**Comment by chullhwan-song** [comment]
- 컴퓨터가 어떤 단어에 대해 인지할 수 있게 하기 위해서는 수치적인 방식으로 단어를 표현
  - problem) 수치화를 통해 단어의 개념적 차이를 나타내기가 힘듬
  - 이 한계를 극복하기 전, 많이 사용하는 방식이 "one-hot encoding"
    - 예) Naive Bayes를 이용한 스팸 분류기에 사용한 방식
    - 나름 좋은 성능을 내었고, 지금까지도 사용
    - 문제) 컴퓨터 자체가 ‘단어가 본질적으로 다른 단어와 어떤 차이점을 가지는 지 이해할 수가 없다’ 는 아주 큰 단점이 존재
  - 벡터화 - one-hot encoding 단점을 해결하기 위해 고안
    - 장점
      - 각 단어들 사이의 유사도를 측정
      - 여러 개의 단어에 대해 다룰 때도 더해서 평균을 내는 등 수치적으로 쉽게 다룰수 있음. 
      - 의미 자체가 벡터로 수치화되어 있기 때문에, 벡터 연산을 통해서 추론을 내릴 수가 있다는 점
        - 추론 데모 : http://w.elnn.kr/search/
    - 방법
      - NNLM, RNNLM, CBOW, Skip-gram 등이 있음.
      - 이를 발전시켜 ->  word2vec
        - Feed-Forward Neural Net Language Model (NNLM) : Neural Network 이용한 방법
![image](https://user-images.githubusercontent.com/40360823/43692102-6226fb52-995f-11e8-8cd5-8fc4892c1d60.png)
        - Input Layer, Projection Layer, Hidden Layer, Output Layer 구성
        - 입력 - N개의 one hot vector,  dictionary size : V -> 1개의 Vector는 V 차원
        - Projection Layer P이면, VxP(Projection Matrix)  -> Hidden Layer 으로 넘기게된다.
          - 실제 역할은? dimension reduction 역할인가?
            - P 는 임의의 수인가?
            - 실제 INPUT 크기(?)
            - 그림에서는 그냥 합치는정도인데.ㅎ 
- word2vec
  - Distributional_Hypothesis 개념하에서 이루어짐
    - 비슷한 분포를 가진 단어들은 비슷한 의미를 가진다란 개념
      - 여기서 비슷한 분포를 가졌다는 것은 기본적으로 단어들이 같은 문맥에서 등장한다는 것을 의미
  - 대표적인 방법 
    - CBOW(Continuous Bag-of-Words) 
![image](https://user-images.githubusercontent.com/40360823/43692104-672bb62e-995f-11e8-994b-6b62b532180f.png)
      - only one word에 대한 CBOW 모델
      - V : vocabulary(dictionary) size
      - N : hidden layer size
      - W : input layer와 hidden layer 사이의 weight = VxN
      - input/hidden/output layer로 구성
        - hidden layer는 projection layer에 더 가까움
          - why?) weight를 곱해주는 layer가 아닌, 단순히 projection 역할에 더 가깝기 때문
        - C -> CxN -> NxV
          - C : 단어
          - N : Hidden layer의 크기
          - V : 사전의 크기
        - 이전방법보다 엄청나게 계산량을 줄임.  
![image](https://user-images.githubusercontent.com/40360823/43692108-6c52ab44-995f-11e8-9408-31b580c08e4e.png)
      - multi-word에 대한 CBOW 모델
        - when computing the hidden layer output, 
          - instead of directly copying the input vector of the input context word, 
            - the CBOW model takes the **average** of the vectors of the input context words
    - Skip-gram
![image](https://user-images.githubusercontent.com/40360823/43692111-70fcfde8-995f-11e8-88b6-036f723b2b71.png)
      - CBOW모델의 반대
      - 현재 주어진 단어 하나를 가지고 주위에 등장하는 나머지 몇 가지의 단어들의 등장 여부를 유추하는 것.
      -  Skip-gram이 CBOW에 비해 전체적으로 다소 좋은 결과를 내는 추세
- ref
  - http://solarisailab.com/archives/374
  - hword2vec 관련 이론 정리 : [링크](https://shuuki4.wordpress.com/2016/01/27/word2vec-%EA%B4%80%EB%A0%A8-%EC%9D%B4%EB%A1%A0-%EC%A0%95%EB%A6%AC/)
  - https://medium.com/@jameschen_78678/social-behavior-analytics-201-523199badba0
